### DISCUSSION TITLE
Has anyone successfully submitted Carton contents and been able to retrieve Labels?

### DISCUSSION THREAD

**Discussion by cpiggott** [amzn/selling-partner-api-models#3190] [General]
I'm reaching out to see if anyone has had any success with getting Unique shipment labels from the API? 

There is an action in the FbaInbound API called [getLabels](https://github.com/amzn/selling-partner-api-docs/blob/a56f4fac467c0e251b9b89ae39ba7c12be65e776/references/fulfillment-inbound-api/fulfillmentInboundV0.md#get-fbainboundv0shipmentsshipmentidlabels) that allows you to retrieve box labels and shipping labels for your shipment.

To do this however, you have to submit the Cart Contents XML document to the Feeds API(Feed type: _POST_FBA_INBOUND_CARTON_CONTENTS_). I've successfully submitted documents to the feed about the shipment, including trying multiple different CartonID's for the cartons with not success on the other side retrieving the shipment labels.

When submitting Carton ID's, I've tried in these different formats:
- Incrementing integer per box(1,2,3,4,5...)
- Incrementing with the "U00000" prefix before each carton(U000001, U000002,...,U000010) which is also accepted
- Incrementing with the "FBA_SHIPMENT_IDU00000" prefix before each for the carton contents.

All of these have been submitted via the Feeds API and are accepted, checking that the feed was accepted, the transport details were submitted and shipment was confirmed.

When I try to retrieve my labels, I am unable to retrieve them via the GetLabels API endpoint. I keep getting an error:

None of these work when supplying the Carton IDs to get labels in the format:
```
[400] {
  "errors": [
    {
      "code": "InvalidInput",
      "message": "Provided SellerCartonIds are not valid: [[1]] for Shipment: [FBAXXXXXXXXX]",
      "details": ""
    }
  ]
}
```

```
[400] {
  "errors": [
    {
      "code": "InvalidInput",
      "message": "Provided SellerCartonIds are not valid: [[U000001]] for Shipment: [FBAXXXXXXXXX]",
      "details": ""
    }
  ]
}
```

```
[400] {
  "errors": [
    {
      "code": "InvalidInput",
      "message": "Provided SellerCartonIds are not valid: [[FBAXXXXXXXXXU000001]] for Shipment: [FBAXXXXXXXXX]",
      "details": ""
    }
  ]
}
```

It's incredibly frustrating and I'm not sure what's going on. Has anyone had any luck with this?

**Comment by CarlosSilesNetrush** [comment]
When you create a shipment in Amazon you need to specify IntendedBoxContentsSource:
Values:
- NONE - There is no box content information for this shipment. Amazon will manually process the box contents. This may incur a fee.
- FEED - Box content information is provided through the _POST_FBA_INBOUND_CARTON_CONTENTS_ feed.
- 2D_BARCODE - Box content information is provided by a barcode on the shipment.

From what I can see, you are trying to do 2D_Barcode format, if that's the case, **you SHOULD NOT submit Carton_Feed**
You just need to Select the Shipping Service First ( Small Parcel or LTL), then just get ALL labels ( do not use unique labels), Amazon will send you the labels with a sequence ( FBAXXXXXU000001, FBAXXXXXU000002)

2D_Barcode generator
```
private static string Get2DBarcodeProductLine(string FNSKU, int qty, DateTime? expiration)
      {
          return $"FNSKU:{FNSKU},QTY:{qty}{(expiration.HasValue ? $",EXP:{expiration.Value:yyMMdd}" : "")}";
      }
```
Here is what you want for your carton:

```
loop through your cartons and call this increasing the int number, this helper will give you the right format

public static string GetCartonIdFor2DBarcode(string shipmentId, int number)
        {           
            var val = (1000000 + number).ToString();
            var valLength = val.Length - 1;
            var plate = $"{shipmentId}U{val.ToString().Substring(1, valLength)}";
            return plate;
        }

```

**If you are trying to use FEED, here is the generator code sample**
We don't do anything with the CartonId , we use an auto-generated int converted to a string. Here is how we create the Feed:

````
public static List<CartonContentsFeed> GenerateCartonContentFeeds(List<string> amzShipmentIds = null)
{               
        var inventory = inventoryQuery.Where(x => )
            .Select(x => new
            {
                x.Container.Order.MarketplaceId,
                ShipmentId = AmazonShipmentIdy,
                x.ContainerId,
                x.NRCode,
                x.Quantity,
                x.ExpirationDate
            })
            .ToList();              

        var cartonContentsRequestsByMarketplace =
            inventory
            .GroupBy(x => x.MarketplaceId).Select(grp => new
            {
                MarketplaceId = grp.Key,
                AmazonEnvelopeMessageCartonContentsRequest = grp.GroupBy(x => x.ShipmentId)
                .Select(grp2 => new AmazonEnvelopeMessageCartonContentsRequest
                {
                    ShipmentId = grp2.Key,
                    NumCartons = grp2.GroupBy(x => x.ContainerId).Count(),
                    Cartons = grp2.GroupBy(x => x.ContainerId).OrderBy(x => x.Key).Select(grp3 => new AmazonEnvelopeMessageCartonContentsRequestCarton
                    {
                        CartonId = grp3.Key.ToString(), // example 100154 ( we do not modify this, this is numeric value originally convert to string)
                        Items = grp3.GroupBy(x => x.NRCode).Select(grp4 => new AmazonEnvelopeMessageCartonContentsRequestCartonItem
                        {
                            SKU = grp4.Key,
                            QuantityShipped = grp4.Sum(x => x.Quantity),
                            ExpirationDate = grp4.Any(x => x.ExpirationDate != null) ? grp4.Select(x => x.ExpirationDate).FirstOrDefault() : null
                        })
                        .ToList()
                    })
                    .ToList()
                })
                .ToList()
            })
            .ToList() 
```

  ```
      var feeds = new List<CartonContentsFeed>();
        cartonContentsRequestsByMarketplace.ForEach(x =>
        {
            var marketplaceId = x.MarketplaceId ?? MarketplaceId.AmazonCOM;
            var creds = YOUR CREDENTIALS;

            var request = new CartonContentsRequestAmazonEnvelope(creds.SellerId);

            x.AmazonEnvelopeMessageCartonContentsRequest.ForEachWithIndex((item, i) =>
            {
                request.Messages.Add(new CartonContentRequestAmazonEnvelopeMessage
                {
                    MessageID = i + 1,
                    CartonContentsRequest = item
                });
            });

            var value = XMLHelpers.SerializeObjectToXML(request);

            feeds.Add(new CartonContentsFeed(marketplaceId, request));
        });

        return feeds;
}`
```

**Reply by cpiggott** [reply]
Hey @CarlosSiles21 ,

I appreciate the detailed response but we actually account for all the different types, Feed, 2D_Barcode, Interactive and that doesn't seem to be the issue. 

I believe I have found my issue and it was with the XML generation. I'm awaiting confirmation that it was the issue from Seller Central support.

The problem surrounding this now is that I submitted "BAD" XML, it was malformed, but it was still processed and considered successfully submitted and accepted by SP-API, even though this wasn't the case after looking through with with ScratchPad.

**Reply by CarlosSilesNetrush** [reply]
You are welcome!
I'm glad you found the problem.
ScratchPad is the best way to find any issue with API.

**Reply by cpiggott** [reply]
@CarlosSiles21 Do you have an example of the XML that you posted that works? 

I've formatted the XML to the exact specifications of the XSD documentation and keep getting Carton was not added, but my XML looks pretty much exactly how it seems to be wanted. I'm posting mine below too, do you see anything glaring that could be an issue?

```
<?xml version="1.0" encoding="UTF-8"?>
<AmazonEnvelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="amzn-envelope.xsd">
	<Header>
		<DocumentVersion>1.02</DocumentVersion>
		<MerchantIdentifier>MERCHANT_ID_REMOVED_FOR_PRIVACY</MerchantIdentifier>
	</Header>
	<MessageType>CartonContentsRequest</MessageType>
	<Message>
		<MessageID>1</MessageID>
		<CartonContentsRequest>
			<ShipmentId>SHIPMENT_STRING_REMOVED_FOR_PRIVACY</ShipmentId>
			<NumCartons>1</NumCartons>
			<Carton>
				<CartonID>000001</CartonID>
				<Item>
					<SKU>SKU_REMOVED_FOR_PRIVACY(String)</SKU>
					<QuantityShipped>6</QuantityShipped>
					<QuantityInCase>1</QuantityInCase>
				</Item>
			</Carton>
		</CartonContentsRequest>
	</Message>
</AmazonEnvelope>
```

**Reply by CarlosSilesNetrush** [reply]
@cpiggott  My only observation is the carton Id starting with 000001, I would try to start with 100001 (maybe that helps), I'm posting my XML below. I hope it helps.

![image](https://user-images.githubusercontent.com/95246651/146049295-89135a40-8df5-48de-a444-6d1f8da5c59c.png)

**Reply by cumaliaba** [reply]
Is there reverse way via SP API to get items list according to each carton(box)?
We are creating shipment through seller central and I want to get items list programmatically via SP API for each box?

Thanks

**Comment by uk-2** [comment]
Happy New year !

Which api call did you use to generate FNSKU labels ?

Also where do we get FNSKU values from @cpiggott @CarlosSiles21

**Reply by CarlosSilesNetrush** [reply]
@uk-2 Happy 2022!  
You don't need to call an API to get the FNSKU, I assumed that you have this value stored in your system upon product setup, you also need to display expiration if it applies, and the condition(new or used). We use 2"x1" inch labels.
